File: app/src/main/java/com/zhai/mvpdemo/view/adapter/NoteAdapter.java
package com.zhai.mvpdemo.view.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.zhai.mvpdemo.R;

import java.util.List;

/**
 * Created by zhai on 16/6/18.
 */

public class NoteAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<String> list;
    private DeleteButtonListener listener;

    public NoteAdapter(List<String> list, DeleteButtonListener listener) {
        this.list = list;
        this.listener = listener;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_item, null, false));
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
        ViewHolder viewHolder = (ViewHolder) holder;
        viewHolder.tv.setText(list.get(position));
        viewHolder.bn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                listener.onDeleteButtonClick(position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return list == null ? 0 : list.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        public TextView tv;
        public Button bn;

        public ViewHolder(View itemView) {
            super(itemView);
            tv = (TextView) itemView.findViewById(R.id.content);
            bn = (Button) itemView.findViewById(R.id.remove);
        }
    }

    public interface DeleteButtonListener {
        void onDeleteButtonClick(int position);
    }
}

File: app/src/main/java/com/zhai/mvpdemo/view/adapter/NoteAdapter.kt
package com.zhai.mvpdemo.view.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView

import com.zhai.mvpdemo.R

/**
 * Created by zhai on 16/6/18.
 */

class NoteAdapter(private val list: List<String>?, private val listener: NoteAdapter.DeleteButtonListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_item, null, false))
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as ViewHolder
        viewHolder.tv.text = list!![position]
        viewHolder.bn.setOnClickListener { listener.onDeleteButtonClick(position) }
    }

    override fun getItemCount(): Int {
        return if (list == null) 0 else list.size
    }

    internal inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var tv: TextView
        var bn: Button

        init {
            tv = itemView.findViewById(R.id.content) as TextView
            bn = itemView.findViewById(R.id.remove) as Button
        }
    }

    interface DeleteButtonListener {
        fun onDeleteButtonClick(position: Int)
    }
}
